[datadog_org_group] Terraform support for org groups and policies (#3686)
DataDog/terraform-provider-datadog: docs/resources/org_group_policy_override.md

resource "datadog_org_group" "prod" {
  name = "Production Environments"
}

resource "datadog_org_group_membership" "exempt_org" {
  org_group_id = datadog_org_group.prod.id
  org_uuid     = "ff4a8255-6931-58d1-add0-a6b3602d5421"
}

resource "datadog_org_group_policy" "widget_copy_paste" {
  org_group_id     = datadog_org_group.prod.id
  policy_name      = "is_widget_copy_paste_enabled"
  content          = jsonencode({ "org_config" : false })
  enforcement_tier = "DEFAULT"
}

# Exempts the given organization from the widget_copy_paste policy.
# The org keeps its current value for is_widget_copy_paste_enabled regardless
# of the policy. The resource must target a policy whose tier is not ENFORCE.
resource "datadog_org_group_policy_override" "example" {
  org_group_id = datadog_org_group.prod.id
  policy_id    = datadog_org_group_policy.widget_copy_paste.id
  org_uuid     = "ff4a8255-6931-58d1-add0-a6b3602d5421"
  org_site     = "us1"
  depends_on   = [datadog_org_group_membership.exempt_org]
}
```

## Behavior notes

### Server-side auto-creation

Overrides are also created automatically by the server when an org's existing config differs from a policy's value. Two triggers:

> 1. **Membership change** — when an org is moved into an org group, the server compares that org's current config to each non-`ENFORCE` policy in the group. Any mismatch produces an auto-created override recording the org's existing value.

Reviewer: ```suggestion
1. **Membership change:** When an org is moved into an org group, the server compares that org's current config to each non-`ENFORCE` policy in the group. Any mismatch produces an auto-created override that records the org's existing value.
```

> 2. **Policy create/update** with a non-`ENFORCE` tier — the server performs the same comparison across every member org at policy apply time, creating overrides for orgs whose config doesn't match the new policy value.

Reviewer: ```suggestion
2. **Policy create/update** with a non-`ENFORCE` tier: The server performs the same comparison across every member org at policy apply time, creating overrides for orgs whose config doesn't match the new policy value.
```

> Auto-created and user-declared overrides are indistinguishable at the API level. They can be adopted into Terraform via `terraform import` or by iterating the [`datadog_org_group_policy_overrides`](../data-sources/org_group_policy_overrides) data source with `for_each` + `import` blocks.

Reviewer: same question about the double periods here

### Delete behavior

> Removing an override does **not** just remove the exemption marker — it resets the target org's config value to match the parent policy's current value. The server treats override deletion as "re-apply the policy to this org" and propagates the policy value accordingly.

Reviewer: ```suggestion
Removing an override does **not** just remove the exemption marker. The target org's config value is also reset to match the parent policy's current value. The server treats override deletion as "re-apply the policy to this org" and propagates the policy value accordingly.
```

If you want to stop managing an override without changing the org's value, use `terraform state rm` instead of removing the resource block.

### ENFORCE tier cascade

> When the parent [`datadog_org_group_policy`](./org_group_policy) transitions to `enforcement_tier = "ENFORCE"`, the server atomically deletes every override for that policy. Creating an override against an already-`ENFORCE`-tier policy also fails with a `FailedPrecondition` error.

Reviewer: just double-checking: this is supposed to say "atomically" and not "automatically" right?

Author: Yes! But I wonder if that is too much detail. I'll change the wording to "automatically" since that was already implied and we don't need to convey the atomic nature of the deletion!

> If you plan to flip a policy to `ENFORCE`, remove the `datadog_org_group_policy_override` resource blocks for that policy in the same commit — otherwise Terraform's next apply will try to re-create the server-deleted overrides and fail.

Reviewer: ```suggestion
If you plan to flip a policy to `ENFORCE`, remove the `datadog_org_group_policy_override` resource blocks for that policy in the same commit. Otherwise, Terraform's next apply will try to re-create the server-deleted overrides and fail.
```

<!-- schema generated by tfplugindocs -->
## Schema

### Required

- `org_group_id` (String) The UUID of the org group that owns the policy. Must be a valid UUID.

> - `org_site` (String) The short site name of the organization (e.g. `us1`, `eu1`, `us1-fed`). Part of the override's server-side identity; changing it replaces the resource. String length must be at least 1.

Reviewer: ```suggestion
- `org_site` (String) The short site name of the organization (e.g., `us1`, `eu1`, `us1-fed`). Part of the override's server-side identity; changing it replaces the resource. String length must be at least 1.
```
